### PR TITLE
Add CAAM DMA driver and update all existing CAAM algorithm v2

### PR DESCRIPTION
**Pull request by clementfaure**
Hello,

Here is the follow-up PR of https://github.com/OP-TEE/optee_os/pull/4404
This new PR now includes the CAAM DMA support for both i.MX and LS platforms.
Please note that CAAM is currently not properly enabled on all LS platforms. This PR is already big enough, I will fix this in another PR right after this one gets merged.

Thanks!

### CODE REVIEW

File: core/drivers/crypto/caam/caam_desc.c [hunk 1]
```diff
@@ -80,7 +84,8 @@ void caam_desc_push(struct caam_inring_entry *in_entry, paddr_t paddr)
 
 paddr_t caam_desc_pop(struct caam_outring_entry *out_entry)
 {
-	const uint32_t *a32 = (const uint32_t *)(&out_entry->desc);
+	const uintptr_t v_desc = (uintptr_t)(&out_entry->desc);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/caam_desc.c:87]: The parenthesis around `(&out_entry->desc)` are reduntant.

File: core/drivers/crypto/caam/utils/utils_dmaobj.c [hunk 2]
```diff
@@ -0,0 +1,1405 @@
...
+ *
+ * @obj CAAM DMA object
+ */
+static size_t get_dma_max_alloc_size(struct caamdmaobj *obj)
+{
+	size_t alloc_size = 0;
+	struct priv_dmaobj *priv = NULL;
+
+	if (!obj)
+		return alloc_size;
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/utils/utils_dmaobj.c:1105]: `return 0` would be more appropriate, even if we can see that `alloc_size` is initialized to `0`.

File: core/drivers/crypto/caam/utils/utils_dmaobj.c [hunk 3]
```diff
@@ -0,0 +1,1405 @@
...
+	if (priv->dmabuf.allocated && max_length > priv->dmabuf.allocated &&
+	    priv->dmabuf.allocated > align)
+		max_length = ROUNDDOWN(priv->dmabuf.allocated, align);
+
+	DMAOBJ_TRACE("Prepare SGT/Buffer to do %zu of %zu", max_length,
+		     *length);
+
+	/* Find the first DMA buffer to start with */
+	TAILQ_FOREACH(entry, &priv->list, link)
+	{
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/utils/utils_dmaobj.c:1295]: Please move this to the end of previous line.

File: core/drivers/crypto/caam/utils/utils_mem.c [hunk 4]
```diff
@@ -53,7 +41,7 @@ static void *mem_alloc(size_t size, uint8_t type)
 	void *ptr = NULL;
 	size_t alloc_size = size;
 
-	MEM_TRACE("alloc %zu bytes of type %" PRIu8, size, type);
+	MEM_TRACE("alloc %zu bytes of type %" PRIu8, size_in, type);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/utils/utils_mem.c:44]: This can't compile.

**Review by jenswi-linaro — COMMENTED** [review]
Comments for `drivers: caam: implement CAAM DMA Object`

File: core/drivers/crypto/caam/hash/caam_hash.c [hunk 5]
```diff
@@ -385,183 +470,98 @@ TEE_Result caam_hash_hmac_update(struct hashctx *ctx, const uint8_t *data,
...
-		 * Set the full data size of the input buffer.
-		 * indata.length has been changed when creating the SGT
-		 * object.
-		 */
-		indata.length = len;
 		HASH_TRACE("Posted %zu of input len %zu made %zu", size_topost,
-			   indata.length, size_inmade);
-		ret = caam_cpy_block_src(&ctx->blockbuf, &indata, size_inmade);
+			   len, size_inmade);
+
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/hash/caam_hash.c:551]: Please keep declarations of variables at the beginning of a block.

File: core/drivers/crypto/caam/hash/caam_hash.c [hunk 6]
```diff
@@ -727,11 +674,19 @@ void caam_hash_hmac_copy_state(struct hashctx *dst, struct hashctx *src)
 
 	assert(dst && src);
 
+	if (!dst->initialized)
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/hash/caam_hash.c:677]: Prefer to write this as:
`if (!dst->initialized && caam_hash_hmac_init(dst))`

**Review by jenswi-linaro — COMMENTED** [review]
Comments for `drivers: caam: use CAAM DMA object in Hash`

File: core/drivers/crypto/caam/cipher/caam_cipher.c [hunk 7]
```diff
@@ -669,157 +568,207 @@ static TEE_Result do_update_streaming(struct drvcrypt_cipher_update *dupdate)
...
+			if (ret)
+				goto end_streaming;
+		}
+
+		/*
+		 * Save the input data in the block buffer for next operation
+		 * and prepare the source DMA Object with the overall saved
+		 * data to generate destination bytes.
+		 */
 		struct caambuf cpysrc = {
```
Comment (jenswi-linaro) [inline core/drivers/crypto/caam/cipher/caam_cipher.c:711]: Please keep declarations of variables at the start of a block.

**Review by jenswi-linaro — COMMENTED** [review]
Comment for `drivers: caam: use CAAM DMA object in Cipher`

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 8]
```diff
@@ -33,11 +39,35 @@ $(call force,CFG_JR_INDEX,2)  # Default JR index used
 
 ifneq (,$(filter $(PLATFORM_FLAVOR),ls1046ardb))
 $(call force,CFG_JR_INT,137)  # Default JR IT Number (105 + 32) = 137
-endif
-
-ifneq (,$(filter $(PLATFORM_FLAVOR),lx2160ardb))
+# Linux indicate the IRQ is 73 + 32 = 105
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-ls/crypto_conf.mk:42]: Leftover comment?

**Review by jenswi-linaro — COMMENTED** [review]
Comment for `core: ls: set SGT properties for LS platforms`

**Comment by jenswi-linaro** [comment]
I've skimmed through with with a few minor comments above. With those addressed:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/drivers/crypto/caam/caam_desc.c (hunk 1 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/caam_desc.c:87]: Ok

File: core/drivers/crypto/caam/utils/utils_dmaobj.c (hunk 2 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c:1105]: OK

File: core/drivers/crypto/caam/utils/utils_dmaobj.c (hunk 3 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c:1295]: I missed this one, I will fix it

File: core/drivers/crypto/caam/utils/utils_mem.c (hunk 4 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_mem.c:44]: Oops, fixed

File: core/drivers/crypto/caam/hash/caam_hash.c (hunk 5 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/hash/caam_hash.c:551]: Ok

File: core/drivers/crypto/caam/hash/caam_hash.c (hunk 6 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/hash/caam_hash.c:677]: No problem

File: core/drivers/crypto/caam/cipher/caam_cipher.c (hunk 7 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/cipher/caam_cipher.c:711]: Ok

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 8 above)
Comment (clementfaure) [reply core/arch/arm/plat-ls/crypto_conf.mk:42]: I will remove these comments

**Comment by clementfaure** [comment]
@jenswi-linaro Thanks for your time. Tag applied

**Comment by sahilnxp** [comment]
For commit `core: ls: set SGT properties for LS platforms` In patch description minor typing error: Set the following SGT properties for `imx` platforms. It should be `LS `in place of `imx`

**Comment by clementfaure** [comment]
@sahilnxp Fixed!

File: core/arch/arm/plat-ls/conf.mk [hunk 9]
```diff
@@ -170,8 +170,7 @@ ifeq ($(CFG_NXP_CAAM),y)
 # If NXP CAAM Driver is supported, the Crypto Driver interfacing
 # it with generic crypto API can be enabled.
 CFG_CRYPTO_DRIVER ?= y
-CFG_CAAM_64BIT ?= y
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/arch/arm/plat-ls/conf.mk:173]: It has been move to ```crypto_conf.mk```

https://github.com/OP-TEE/optee_os/pull/4481/commits/44890b777b7fe2a4dfcfc64694ee7770534194e1#diff-05cfcdcbc667074168e718479474feec228e38c2558fe097ba2d30a376662a32R26

**Comment by sahilnxp** [comment]
`Acked-by: Sahil Malhotra <sahil.malhotra@nxp.com>`

**Comment by clementfaure** [comment]
Force-push:
- Apply tag from @sahilnxp
- Fix copyright updates for non-consecutive years.
```Copyright 2018-2019 NXP``` becomes ```Copyright 2018-2019, 2021 NXP``` and not ```Copyright 2018-2021 NXP```

**Comment by jforissier** [comment]
>     * Apply tag from @sahilnxp

Shouldn't it be applied to all commits?

**Comment by clementfaure** [comment]
@jforissier I (wrongfully ?) assume this tag was only for ```core: ls: set SGT properties for LS platforms``` since the only comment made was for this commit.
@sahilnxp Does your tag apply to all commits or ```core: ls: set SGT properties for LS platforms``` only ?

**Comment by sahilnxp** [comment]
@clementfaure @jforissier Sorry for creating confusion.
My tag applies to `core: ls: set SGT properties for LS platforms only.`

**Comment by jforissier** [comment]
@sahilnxp no worries, I suspected it was the case actually ;-)
Thanks all, I am merging this now.

**Pull request merged**